src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -553,10 +565,69 @@ static void on_resolver_result_changed_locked(void* arg, grpc_error* error) {
       // Instantiate new LB policy.
       create_new_lb_policy_locked(chand, lb_policy_name.get(),
                                   &connectivity_state, &connectivity_error);
+      if (chand->channelz_channel != nullptr) {
+        char* str;
+        gpr_asprintf(&str, "Switched LB policy to %s", lb_policy_name.get());
+        trace_strings.push_back(str);
+      }
     }
     // Find service config.
     grpc_core::UniquePtr<char> service_config_json =
         get_service_config_from_resolver_result_locked(chand);
+    // Note: It's safe to use chand->info_service_config_json here without
+    // taking a lock on chand->info_mu, because this function is the
+    // only thing that modifies its value, and it can only be invoked
+    // once at any given time.
+    if (chand->channelz_channel != nullptr) {
+      if ((service_config_json == nullptr &&
+           chand->info_service_config_json != nullptr) ||
+          (service_config_json != nullptr &&
+           chand->info_service_config_json == nullptr) ||
+          (service_config_json != nullptr &&
+           chand->info_service_config_json != nullptr &&
+           strcmp(service_config_json.get(),
+                  chand->info_service_config_json.get()) != 0)) {
+        // TODO(ncteisen): might be worth somehow including a snippet of the
+        // config in the trace, at the risk of bloating the trace logs.
+        trace_strings.push_back(gpr_strdup("Service config reloaded"));
```

Comment:
The phraseology here can be a bit misleading; in cases like the tests where we inject the service config via the client API, there's no "reload" going on.  I suggest changing this to just say "service config changed".

Reply:
Done